src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp

```diff
@@ -0,0 +1,1274 @@
+                instance->DoRemoveAurasDueToSpellOnPlayers(SPELL_BLAZE_OF_GLORY);
+                firelands_bossAI::JustDied(killer);
+            }
+
+            void EnterEvadeMode(EvadeReason reason) override
+            {
+                instance->DoRemoveAurasDueToSpellOnPlayers(SPELL_BLAZE_OF_GLORY);
+                firelands_bossAI::EnterEvadeMode(reason);
+            }
+
+            void UpdateAI(uint32 diff) override
+            {
+                scheduler.Update(diff);
+                separateScheduler.Update(diff);
+                firelands_bossAI::UpdateAI(diff);
+            }
+
+            void SetGUID(ObjectGuid guid, int32 type = 0) override
+            {
+                switch (type)
+                {
+                    case GUID_TORMENTED:
+                        for (int i = 0; i < 25; i++)
+                        {
+                            if (_sharedThePain[i].player == guid)
+                            {
+                                _sharedThePain[i].tormented += 1;
+                                break;
+                            }
+                            if (_sharedThePain[i].player.IsEmpty())
+                            {
+                                _sharedThePain[i].player = guid;
+                                _sharedThePain[i].tormented += 1;
+                                break;
+                            }
+                        }
+                        break;
+                    default:
+                        break;
+                }
+            }
+
+            bool SharedThePain() const
+            {
+                for (int i = 0; i < 25; i++)
+                    if (_sharedThePain[i].tormented > 3)
+                        return false;
+
+                return true;
+            }
+
+            private:
+                void EquipWeapon(uint8 equipment) const
+                {
+                    switch (equipment)
+                    {
+                        case EQUIP_DEFAULT:
+                            me->LoadEquipment(equipment);
+                            me->SetCanDualWield(true);
+                            break;
+                        case EQUIP_INFERNO_BLADE:
+                        case EQUIP_DECIMATION_BLADE:
+                            me->LoadEquipment(equipment);
+                            me->SetCanDualWield(false);
+                            break;
+                        default:
+                            break;
+                    }
+                }
+                // Our default TaskScheduler has a UNIT_STATE_CASTING validator that would get in the way of certain tasks, run them on a separate track.
+                TaskScheduler separateScheduler;
+                bool _canYellKilledPlayer;
+                struct _sharedThePain
+                {
+                    ObjectGuid player;
+                    uint32 tormented;
+                }
+                _sharedThePain[25];
```

Comment:
always default to std::unordered_map unless for some reason you need it sorted by key